Support for enzop vlct

mabruzzo opened the pull request:
<!--Thank you so much for your PR! To help us review, fill out the form
to the best of your ability.  Please make use of the development guide at
http://yt-project.org/docs/dev/developing/index.html-->

<!--Provide a general summary of your changes in the title above, for
example "Raises ValueError on Non-Numeric Input to set_xlim".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

<!--If you are able to do so, please do not create the PR out of main, but out
of a separate branch. -->

## PR Summary

This PR introduces support for outputs generated by Enzo-E's VL+CT integrator. This basically just adds handling for magnetic fields. It also adds proper handling for determining `"specific_thermal_energy"` when the dual-energy-formalism is being used.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

I just wanted a little clarification on the tests I should be adding. Would it be adequate to introduce an answer test involving a new dataset generated with this integrator and an answer test similar to the `test_hello_world` in `enzo_p/tests/test_outputs.py`? Should I also be introducing an additional dataset for checking the dual-energy-formalism handling? 

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->
- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or the
recommended next step seems overly demanding , or if you would like help in
addressing a reviewer's comments.  And please ping us if you've been waiting
too long to hear back on your PR.-->

chummels commented:
Hi @mabruzzo !  I just issued PR #3273  to swap enzo-p references in for enzo-e in response to some confusion by new enzo-e users.  I imagine it will yield a bunch of merge conflicts depending on when our respective PRs go through--probably not too gnarly, but just wanted to make sure you were aware of it.

mabruzzo commented:
Hi, @chummels . Thanks for pointing that out. I'm really happy to see those changes!

neutrinoceros commented:
@chummels would you advise that we give priority to this over https://github.com/yt-project/yt/pull/3273 ?
It seems to me it's probably easier to solve conflicts there rather than here.

chummels commented:
I figured mine was basically ready to go, and is in direct response to the enzo-e dev meeting that is happening this week.  I don't know how much modification this PR needs or if there will be other relevant PRs for enzo-e this week.

mabruzzo commented:
@chummels, neutrinoceros,  For what it's worth, I would encourage you to give priority to PR #3273 . It might take a few days before I have a chance to address everything that is blocking this from going through.

neutrinoceros commented:
Good to know, thank you !

neutrinoceros commented on yt/frontends/enzo_p/fields.py:
```diff
@@ -40,14 +53,71 @@ class EnzoPFieldInfo(FieldInfoContainer):
+        else:
+
+            def _tot_minus_kin(field, data):
+                ret = data[("enzop", "total_energy")] - 0.5 * data[("gas", "velocity_x")] ** 2.0
+                if data.ds.dimensionality > 1:
+                    ret -= 0.5 * data[("gas", "velocity_y")] ** 2.0
+                if data.ds.dimensionality > 2:
+                    ret -= 0.5 * data[("gas", "velocity_z")] ** 2.0
+                return ret
```
quick suggestion
```suggestion
                dim =  data.ds.dimensionality
                return data[("enzop", "total_energy")] - dim * 0.5 * data[("gas", "velocity_x")] ** 2.0
```

neutrinoceros commented on yt/frontends/enzo_p/fields.py:
```diff
@@ -40,14 +53,71 @@ class EnzoPFieldInfo(FieldInfoContainer):
+        else:
+
+            def _tot_minus_kin(field, data):
+                ret = data[("enzop", "total_energy")] - 0.5 * data[("gas", "velocity_x")] ** 2.0
+                if data.ds.dimensionality > 1:
+                    ret -= 0.5 * data[("gas", "velocity_y")] ** 2.0
+                if data.ds.dimensionality > 2:
+                    ret -= 0.5 * data[("gas", "velocity_z")] ** 2.0
+                return ret
```
a quick suggestion
```suggestion
                return data["enzop", "total_energy"] - 0.5 * data["gas", "velocity_magnitude"] ** 2.0
```

neutrinoceros replied:
note that parens are not necessary in data objects' __gettiem__ method :)

neutrinoceros commented:
@mabruzzo Sure enough now that  #3273 was merged, you're getting conflicts here. Don't hesitate to ping me or Cameron if you need help to salvage this branch !

mabruzzo commented:
I rebased this branch off of the main branch to resolve the merge conflict.  

I also introduced a new answer test to `yt/frontends/enzo_e/tests/test_outputs.py` to validate that the that face-centered magnetic fields are read in properly (this mirrors a similar test present in the Enzo frontend). This new answer test requires a new sample dataset: [orszag-tang_0.5.tar.gz](https://github.com/yt-project/yt/files/6518922/orszag-tang_0.5.tar.gz) (@brittonsmith this is really similar to the test dataset that I emailed you for the Enzo-E workshop, but I reran it tonight after the main branch was updated and the output actually includes the interface magnetic field data)

A couple of questions:

- Would you mind telling me if there are any other steps I need to take to introduce that test dataset?
- Should I also be introducing another dataset to write some explicit answer tests for the dual-energy formalism handling? Or do you think this is good enough?
- Should I be updating any documentation? I suppose that I could add a blurb like the MHDCT section for the Enzo frontend, but I'm not really sure that people are actually going to load in the face-centered magnetic field values (aside for debugging purposes). If they are going to look at the magnetic field for science purposes, I suspect that they will probably just use the cell-centered values.

Finally, I wanted to note that when I run the Enzo-E answer tests locally, I got the following error message:

<details>
  <summary>Error Message</summary>
  
  ```bash
.............................................................E....
======================================================================
ERROR: Failure: SymbolNotFoundError (The symbol 'code_velocity' does not exist in this registry.)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/unyt/unit_registry.py", line 72, in __getitem__
    ret = self.lut[str(key)]
KeyError: 'code_velocity'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/unyt/unit_registry.py", line 75, in __getitem__
    _lookup_unit_symbol(str(key), self.lut)
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/unyt/unit_registry.py", line 331, in _lookup_unit_symbol
    "Could not find unit symbol '%s' in the provided " "symbols." % symbol_str
unyt.exceptions.UnitParseError: Could not find unit symbol 'code_velocity' in the provided symbols.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/nose/loader.py", line 252, in generate
    for test in g():
  File "/home/mabruzzo/packages/py3/yt/yt/frontends/enzo_e/tests/test_outputs.py", line 67, in test_particle_fields
    yield FieldValuesTest(ep_cosmo, field, dobj_name, particle_type=True)
  File "/home/mabruzzo/packages/py3/yt/yt/utilities/answer_testing/framework.py", line 495, in __init__
    super().__init__(ds_fn)
  File "/home/mabruzzo/packages/py3/yt/yt/utilities/answer_testing/framework.py", line 386, in __init__
    self.ds = data_dir_load(ds_fn, kwargs={"unit_system": "code"})
  File "/home/mabruzzo/packages/py3/yt/yt/utilities/answer_testing/framework.py", line 357, in data_dir_load
    ds.index
  File "/home/mabruzzo/packages/py3/yt/yt/data_objects/static_output.py", line 530, in index
    self.create_field_info()
  File "/home/mabruzzo/packages/py3/yt/yt/data_objects/static_output.py", line 604, in create_field_info
    self.field_info.load_all_plugins(self.default_fluid_type)
  File "/home/mabruzzo/packages/py3/yt/yt/fields/field_info_container.py", line 403, in load_all_plugins
    self.find_dependencies(loaded)
  File "/home/mabruzzo/packages/py3/yt/yt/fields/field_info_container.py", line 416, in find_dependencies
    deps, unavailable = self.check_derived_fields(loaded)
  File "/home/mabruzzo/packages/py3/yt/yt/fields/field_info_container.py", line 613, in check_derived_fields
    fd = fi.get_dependencies(ds=self.ds)
  File "/home/mabruzzo/packages/py3/yt/yt/fields/derived_field.py", line 256, in get_dependencies
    e[self.name]
  File "/home/mabruzzo/packages/py3/yt/yt/fields/field_detector.py", line 125, in __missing__
    vv = finfo(self)
  File "/home/mabruzzo/packages/py3/yt/yt/fields/derived_field.py", line 297, in __call__
    dd = self._function(self, data)
  File "/home/mabruzzo/packages/py3/yt/yt/fields/cosmology_fields.py", line 154, in _weak_lensing_convergence
    * data[ftype, "matter_overdensity"]
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/unyt/array.py", line 1751, in __array_ufunc__
    mul, unit = unit_operator(u0, u1)
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/unyt/array.py", line 183, in _divide_units
    ret = (unit1 / unit2).simplify()
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/unyt/unit_object.py", line 737, in simplify
    self.expr = _cancel_mul(expr, self.registry)
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/unyt/unit_object.py", line 772, in _cancel_mul
    u1 = _create_unit_from_factor(pair[0], registry)
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/unyt/unit_object.py", line 761, in _create_unit_from_factor
    f = registry[str(base)]
  File "/home/mabruzzo/.pyenv/versions/py37_new/lib/python3.7/site-packages/unyt/unit_registry.py", line 79, in __getitem__
    "The symbol '%s' does not exist in this registry." % key
unyt.exceptions.SymbolNotFoundError: The symbol 'code_velocity' does not exist in this registry.

----------------------------------------------------------------------
Ran 66 tests in 101.095s

FAILED (errors=1)
  ```
  
</details>

This message appears regardless of whether I use the main branch or this feature branch.

neutrinoceros commented:
> Would you mind telling me if there are any other steps I need to take to introduce that test dataset?

There are:
- you'll need to submit a PR to our website repo (see the most recent example https://github.com/yt-project/website/pull/96)
- we will also need to register it in the main repo in `yt/sample_data_registry.json`  (this can be done later)

> Finally, I wanted to note that when I run the Enzo-E answer tests locally, I got the following error message:
> ```The symbol 'code_velocity' does not exist in this registry.```

This is expected. You'll need to replace `"code_velocity"` with `"code_length / code_time"` where due :)

mabruzzo commented:
Should I wait until after the website is updated with the new sample dataset to update:
- `yt/sample_data_registry.json`
- `tests/tests.yaml` with the new answer test

> This is expected. You'll need to replace "code_velocity" with "code_length / code_time" where due :)
I tried to resolve that failing test this morning and wasn't successful.  I tried replacing `"code_velocity"` with `"code_length / code_time"` throughout the enzo-e frontend (the only occurrences are on lines 7-9 of `yt/frontends/enzo_e/fields.py`), but the problem persisted.

I'm pretty sure that this issue relates to loading particles. I was able reproduce this issue by running the following script (which is a trimmed version of `test_ouputs.py`

<details>
  <summary> test.py </summary>

   ```python  
import yt
from yt.testing import assert_array_equal, assert_equal, requires_file
from yt.utilities.answer_testing.framework import (
    FieldValuesTest,
    )
ep_cosmo = "yt-data/ENZOP_DD0140/ENZOP_DD0140.block_list"

_pfields = (
    ("all", "particle_position_x"),
    ("all", "particle_position_y"),
    ("all", "particle_position_z"),
    ("all", "particle_velocity_x"),
    ("all", "particle_velocity_y"),
    ("all", "particle_velocity_z"),
)

def test_particle_fields():
    ds = yt.load(ep_cosmo)

    dso = [None, ("sphere", ("max", (0.1, "unitary")))]
    for dobj_name in dso:
        for field in _pfields:
            yield FieldValuesTest(ep_cosmo, field, dobj_name, particle_type=True)
        dobj = create_obj(ds, dobj_name)
        s1 = dobj[("index", "ones")].sum()
        s2 = sum(mask.sum() for block, mask in dobj.blocks)
        assert_equal(s1, s2)

for elem in test_particle_fields():
    print(elem.run())
   ```


  </details>

I'm not confident that I'm personally going to be able resolve this problem without investing a lot of time (I don't really know anything about how yt handles particles). Because this test failure concerns particles and predates this PR (i.e. you can reproduce the error on the main branch), would it be alright if I deferred this problem to the future and opened an Issue to document it?

If this is all fine, then I _think_ this PR is ready for review.

mabruzzo commented:
So it turns out that I don't need to modify the `tests/tests.yaml` file (the enzo test that my new answer test is based on doesn't have an entry in `tests/tests.yaml` either).

I'm just a little unclear about what the `hash` entry should be set to in `yt/sample_data_registry.json`

neutrinoceros commented:
one quick remark: you need to register the new dataset here to. Specifically in `yt/sample_data_registry.json`
To generate the hash, run
```
openssl sha256 <your tarball>
```
If you need any further help on this, don't hesitate ton ping me

mabruzzo commented:
@neutrinoceros, sorry for bothering you. Do you have an idea why the `py38` and `py38_unittests` checks are failing?

The error seems to say that they can't find `ENZOE_orszag-tang_0.5/ENZOE_orszag-tang_0.5.block_list`. I've looked through my file changes and can't seem to find any obvious problems. It's weird that all of the other checks seem to pass. My only thought is: could this be related to whether I'm using the `@requires_file` decorator instead of the `@requires_ds` decorator? (I'm not sure what the difference is)

Also, do you know why the Mergeable check is failing?

neutrinoceros commented:
> @neutrinoceros, sorry for bothering you. Do you have an idea why the py38 and py38_unittests checks are failing?

no problem ! Looks like the files are either not installed properly on the server or the way we're looking for them is broken
@Xarthisius might know better, but I'll take a moment to inspect this more thoroughly later today.

> It's weird that all of the other checks seem to pass.

No other check is actually using this file, so it's expected :)

>My only thought is: could this be related to using the @requires_file decorator instead of the @requires_ds decorator? (I'm not sure what the difference is, but their implementations seem very similar)

yeah there's a good chance that it's where the problem lie. I hope it's not too urgent, I must run now, but I'll be back !

> Also, do you know why the Mergeable check is failing?

this bot _demands_ we use at least one label from a subset. Not your fault, I'll take care of this now.

Xarthisius commented:
@yt-fido test this please

Xarthisius commented:
> My only thought is: could this be related to using the `@requires_file` decorator instead of the `@requires_ds` decorator? (I'm not sure what the difference is, but their implementations seem very similar)

Your example dataset was not uploaded to testing infrastructure, hence failures. `@requires_ds` is used for answer tests, i.e. special kind of tests where we store results and compare them with current run. `@requires_file` is only used to skip the test if file doesn't exist. Looking at your test you should be using `requires_file` which you are doing.

mabruzzo commented:
Thanks for the help! All of the checks pass now!

neutrinoceros commented:
thank you @Xarthisius !

neutrinoceros commented on yt/frontends/enzo_e/fields.py:
```diff
@@ -40,14 +55,76 @@ class EnzoEFieldInfo(FieldInfoContainer):
     def __init__(self, ds, field_list, slice_info=None):
         super().__init__(ds, field_list, slice_info=slice_info)
 
+        # setup nodal flag information
+        for field in NODAL_FLAGS:
+            if ("enzoe", field) in self:
+                finfo = self["enzoe", field]
+                finfo.nodal_flag = np.array(NODAL_FLAGS[field])
```
```suggestion
        # setup nodal flag information
        for field, arr in NODAL_FLAGS.items():
            if ("enzoe", field) in self:
                finfo = self["enzoe", field]
                finfo.nodal_flag = np.array(arr)
```

neutrinoceros commented on yt/frontends/enzo_e/fields.py:
```diff
@@ -40,14 +55,76 @@ class EnzoEFieldInfo(FieldInfoContainer):
+
     def setup_fluid_fields(self):
-        super().setup_fluid_fields()
+        from yt.fields.magnetic_field import setup_magnetic_field_aliases
```
Unless there's a specific reason (like circular imports ?) this import statement should live at the top of the module

neutrinoceros commented on yt/frontends/enzo_e/fields.py:
```diff
@@ -40,14 +55,76 @@ class EnzoEFieldInfo(FieldInfoContainer):
+        from yt.fields.magnetic_field import setup_magnetic_field_aliases
+
+        self.setup_energy_field()
+        setup_magnetic_field_aliases(self, "enzoe", ["bfield_%s" % ax for ax in "xyz"])
```
```suggestion
        setup_magnetic_field_aliases(self, "enzoe", [f"bfield_{ax}" ax for ax in "xyz"])
```

neutrinoceros commented on yt/frontends/enzo_e/fields.py:
```diff
@@ -40,14 +55,76 @@ class EnzoEFieldInfo(FieldInfoContainer):
+                )
+            else:
+                # thermal energy = total energy - kinetic energy - magnetic energy
+                def _sub_b(field, data):
+                    ret = _tot_minus_kin(field, data)
+                    ret -= (
+                        data["gas", "magnetic_energy_density"] / data["gas", "density"]
+                    )
+                    return ret
```
```suggestion
                def _sub_b(field, data):
                    return _tot_minus_kin(field, data) - data["gas", "magnetic_energy_density"] / data["gas", "density"]
```

neutrinoceros commented on yt/frontends/enzo_e/tests/test_outputs.py:
```diff
@@ -103,3 +105,27 @@ def test_critical_density():
+            assert_equal(grid[field].shape, block_dims + (2 * sum(flag),))
+
+    # Average of face-centered fields should be the same as cell-centered field
+    assert (ad["enzoe", "bfieldi_x"].sum(axis=-1) / 2 == ad["enzoe", "bfield_x"]).all()
+    assert (ad["enzoe", "bfieldi_y"].sum(axis=-1) / 2 == ad["enzoe", "bfield_y"]).all()
+    assert (ad["enzoe", "bfieldi_z"].sum(axis=-1) / 2 == ad["enzoe", "bfield_z"]).all()
```
specialised array comparison methods from numpy produce much more informative messages in case of failure :)
```suggestion
    assert_array_equal(ad["enzoe", "bfieldi_x"].sum(axis=-1) / 2, ad["enzoe", "bfield_x"])
    assert_array_equal(ad["enzoe", "bfieldi_y"].sum(axis=-1) / 2, ad["enzoe", "bfield_y"])
    assert_array_equal(ad["enzoe", "bfieldi_z"].sum(axis=-1) / 2, ad["enzoe", "bfield_z"])
```

neutrinoceros reviewed (changes requested):
Looks good ! thank you for doing this ! I have a few suggestions, the most important of which is for the test you wrote. When they're adressed, I'll be happy to approve and merge

mabruzzo commented:
@neutrinoceros I made the changes (although pre-commit slightly changed the formatting)

neutrinoceros commented:
>  pre-commit slightly changed the formatting

well that's its job, so hurray !
